crates/lang/codegen/src/generator/events.rs

```diff
@@ -236,16 +235,19 @@ impl<'a> Events<'a> {
                         .ident()
                         .map(quote::ToTokens::into_token_stream)
                         .unwrap_or_else(|| quote_spanned!(span => #n));
-                    let field_type = topic_field.ty();
+                    let signature = syn::LitByteStr::new(
+                        format!("{}::{}::{}", contract_ident, event_ident,
+                            field_ident
+                        ).as_bytes(), span);
                     quote_spanned!(span =>
-                        .push_topic::<#field_type>(&self.#field_ident)
+                        .push_topic(&::ink_env::topics::PrefixedValue { value: &self.#field_ident, prefix: #signature })
                     )
                 });
             // Only include topic for event signature in case of non-anonymous event.
             let event_signature_topic = match event.anonymous {
                 true => None,
                 false => Some(quote_spanned!(span=>
-                    .push_topic::<[u8; #len_event_signature]>(#event_signature)
+                    .push_topic(&::ink_env::topics::PrefixedValue { value: #event_signature, prefix: b"" })
```

Comment:
Please correct me if I am wrong but maybe we could have the `#event_signature` as `prefix` since it actually is its prefix while having no real value. Maybe this is confusing though but it would unify the approach taken for fields and their values where basically the same information is the `prefix`.

Reply:
It's a good idea. I tried it out though and stumbled upon that we have `pub prefix: &'a [u8]` and `pub value: &'b T`, so with `value: b""` we then run into `the trait `Encode` is not implemented for `PrefixedValue<'_, '_, [u8; 0]>``.